include/libcyphal/application/registry/registry_impl.hpp

```diff
@@ -0,0 +1,215 @@
+    }
+
+    cetl::optional<SetError> set(const IRegister::Name name, const Value& new_value) override
+    {
+        if (auto* const reg = findRegisterBy(name))
+        {
+            return reg->set(new_value);
+        }
+        return SetError::Existence;
+    }
+
+    // MARK: - IIntrospectableRegistry
+
+    std::size_t size() const override
+    {
+        return registers_tree_.size();
+    }
+
+    IRegister::Name index(const std::size_t index) const override
+    {
+        if (const auto* const reg = registers_tree_[index])
+        {
+            return reg->getName();
+        }
+        return IRegister::Name{};
+    }
+
+    bool append(IRegister& reg) override
+    {
+        CETL_DEBUG_ASSERT(!reg.isLinked(), "Should not be linked yet.");
+
+        auto register_existing = registers_tree_.search(
+            [key = reg.getKey()](const IRegister& other) {
+                //
+                return other.compareBy(key);
+            },
+            [&reg]() -> IRegister* {
+                //
+                return &reg;
+            });
+
+        CETL_DEBUG_ASSERT(std::get<0>(register_existing) != nullptr, "");
+        CETL_DEBUG_ASSERT(std::get<0>(register_existing)->isLinked(), "Should be linked.");
+        return !std::get<1>(register_existing);
+    }
+
+    // MARK: - Other factory methods:
+
+    /// Constructs a new read-only register, and links it to this registry.
+    ///
+    /// @param name The name of the register. Should be unique within the registry.
+    /// @param options Extra options for the register, like "persistent" option (`true` by default).
+    /// @param getter The getter function to provide the register value (from a storage).
+    /// @return Result register if it was appended successfully. Otherwise, `nullopt`.
+    ///
+    template <typename Getter>
+    cetl::optional<Register<Getter, void>> route(const IRegister::Name     name,
+                                                 const IRegister::Options& options,
+                                                 Getter&&                  getter)
+    {
+        auto reg = makeRegister(memory(), name, options, std::forward<Getter>(getter));
+        if (append(reg))
+        {
+            return reg;
+        }
+        return cetl::nullopt;
+    }
+
+    /// Constructs a new read-write register, and links it to this registry.
+    ///
+    /// @param name The name of the register. Should be unique within the registry.
+    /// @param options Extra options for the register, like "persistent" option (`true` by default).
+    /// @param getter The getter function to provide the register value (from a storage).
+    /// @param setter The setter function to update the register value (at the storage).
+    /// @return Result register if it was appended successfully. Otherwise, `nullopt`.
+    ///
+    template <typename Getter, typename Setter>
+    cetl::optional<Register<Getter, Setter>> route(const IRegister::Name     name,
```

Comment:
Can we please also keep the simple route overload that takes only the name and the getter? Read-only non-persistent registers are very common so the helper is, well, helpful.

    /// Convenience method for creating getters with default options.
    template <NameConcept N, Getter G>
    [[nodiscard]] RegisterPtr route(N&& name, G&& getter)
    {
        return add(std::forward<N>(name), Options{}, std::forward<G>(getter), std::monostate{});
    }

Reply:
What do you think about changing order of `options` and `getter`/`setter` lambdas. Like:
```c++
    template <typename Getter>
    cetl::optional<Register<Getter, void>> route(const Name name,
                                                 Getter&& getter,
                                                 const IRegister::Options& options = {});
    template <typename Getter, typename Setter>
    cetl::optional<Register<Getter, void>> route(const Name name,
                                                 Getter&& getter,
                                                 Setter&& setter,
                                                 const IRegister::Options& options = {});
```
In this way we can have `options` parameter with `{}` default? In Swift language it's actually very recommended to have closures (aka lambdas) as last parameters, but it is mostly b/c of very nice "terminal closure" syntax Swift has... But for C++ I'm not aware of any recommendation about the order.

Reply:
Sounds good.